packages/mongodb-chatbot-server/src/routes/generateResponse.ts

```diff
@@ -0,0 +1,342 @@
+      reqId,
+      llm,
+      llmConversation,
+      llmNotWorkingMessage,
+      conversation,
+      noRelevantContentMessage,
+      request,
+    });
+  }
+}
+
+export type AwaitGenerateResponseParams = Omit<
+  GenerateResponseParams,
+  "shouldStream" | "dataStreamer"
+>;
+
+export async function awaitGenerateResponse({
+  reqId,
+  llmConversation,
+  llm,
+  conversation,
+  references,
+  llmNotWorkingMessage,
+  noRelevantContentMessage,
+  request,
+}: AwaitGenerateResponseParams): Promise<GenerateResponseReturnValue> {
+  const newMessages: SomeMessage[] = [];
+  const outputReferences: References = [];
+
+  if (references) {
+    outputReferences.push(...references);
+  }
+
+  try {
+    logRequest({
+      reqId,
+      message: `All messages for LLM: ${JSON.stringify(llmConversation)}`,
+    });
+    const answer = await llm.answerQuestionAwaited({
+      messages: llmConversation,
+    });
+    newMessages.push(convertMessageFromLlmToDb(answer));
+
+    // LLM responds with tool call
+    if (answer?.functionCall) {
+      assert(
+        llm.callTool,
+        "You must implement the callTool() method on your ChatLlm to access this code."
+      );
+      const toolAnswer = await llm.callTool({
+        messages: [...llmConversation, ...newMessages],
+        conversation,
+        request,
+      });
+      logRequest({
+        reqId,
+        message: `LLM tool call: ${JSON.stringify(toolAnswer)}`,
+      });
+      const {
+        toolCallMessage,
+        references: toolReferences,
+        rejectUserQuery,
+      } = toolAnswer;
+      newMessages.push(convertMessageFromLlmToDb(toolCallMessage));
+      // Update references from tool call
+      if (toolReferences) {
+        outputReferences.push(...toolReferences);
+      }
+      // Return static response if query rejected
+      if (rejectUserQuery) {
+        newMessages.push({
+          role: "assistant",
+          content: noRelevantContentMessage,
+        });
+      } // Otherwise respond with LLM again
+      else {
+        const answer = await llm.answerQuestionAwaited({
+          messages: [...llmConversation, ...newMessages],
+          // Only allow 1 tool call per user message.
+          toolCallOptions: "none",
+        });
+        newMessages.push(convertMessageFromLlmToDb(answer));
+      }
+    }
+  } catch (err) {
+    const errorMessage =
+      err instanceof Error ? err.message : JSON.stringify(err);
+    logRequest({
+      reqId,
+      message: `LLM error: ${errorMessage}`,
+      type: "error",
+    });
+    logRequest({
+      reqId,
+      message: "Only sending vector search results to user",
+    });
+    const llmNotWorkingResponse = {
+      role: "assistant",
+      content: llmNotWorkingMessage,
+      references,
+    } satisfies AssistantMessage;
+    newMessages.push(llmNotWorkingResponse);
+  }
+  // Add references to the last assistant message
+  (newMessages[newMessages.length - 1] as AssistantMessage).references =
```

Comment:
I would `assert(newMessages.length > 0)` as any other case would be a programmer error...